### PR TITLE
check if index policy has changed on repo initialization and update if needed

### PR DESCRIPTION
**Pull request by Blackbaud-EricSlater**
This is a fix for https://github.com/Azure/azure-sdk-for-java/issues/17536

### CODE REVIEW

**Comment by kushagraThapar** [comment]
Thanks @Blackbaud-EricSlater for the contribution, we will review them soon.

File: sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/SimpleCosmosRepository.java
Comment (Blackbaud-EricSlater) [inline sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/SimpleCosmosRepository.java:58]: i wanted to go ahead and open this PR to start discussion of this line. The issue here is that the indexPolicy object retrieved from the server will not ever compare with the index policy that is contained in CosmosEntityInformation using .equals(). This is for a few reasons....
1.) The IndexPolicy object only contains a few instance fields and the rest of the data is contained in the serialized version of the object that is contained internally. The version of indexPolicy that is within CosmosEntityInformation will never contain a serialized version of the object that is comparable to the one retrieved from the server due to all the metadata fields that are added to the serialized json by the server.
2.) The version of the indexPolicy object in CosmosEntityInformation is created from the parameters provided by the annotation. However the annotation does not account for any of the implied state that is set in certain cases. For instance, if you apply the CosmosIndexingPolicy annotation with no parameters, then the default indexing policy will get applied to your container. The index policy object in CosmosEntityInformation will contain an empty list for the included and excluded fields in this case. The index policy object retrieved from the server will contain an included list with "/*"  and an excluded list with "/_etag/*".

Since the way the index policy is updated is by doing a container replace, it seems like this check needs to be accurate as far as detecting if the index policy actually changed so that needless container replaces are not being done.

The only approach i can think of to do this is to account for all the discrepancies between these objects when comparing.

Can you guys think of a better way of determining if an index policy update is needed?

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java
Comment (Blackbaud-EricSlater) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java:93]: This was necessary because the container replace method that is being used in azure-cosmos depends on the __self link being in the propertyBag. You can see below in SimpleCosmosRepository that I have to retrieve the CosmosContainerProperties from the server so that this field is contained in the properties bag. I cannot use the one that i create myself using CosmosEntityInformation since it will not contain the __self link. 
So, I retrieve CosmosContainerProperties from the server and then set the new indexing policy on it. This change makes it so that calling setIndexingPolicy updates that underlying json representation that is sent to the server. Without this is place, calling setIndexingPolicy() has no effect.

**Comment by kushagraThapar** [comment]
@moderakh  - please take a look at the changes from azure-cosmos SDK perspective.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java
Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java:93]: @kushagraThapar  @Blackbaud-EricSlater 
There is a test in `com.azure.cosmos.CosmosContainerTest.replace` validating replace scenario. Does this mean that test is broken?

Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java]: im not sure what i missed in my testing that made me think this was necessary, but it is working without it. The DocumentCollection is re-serialized when getV2Collection() is called on the DocumentCollection in CosmosAsyncContainer.replaceInternal. I removed the edits to this file.

However, on a separate note, in looking at the test you mentioned, i did notice that it was not actually changing the index policy to a new value in the test. It was creating a container with indexMode.CONSISTENT and replacing it with a container with indexMode.CONSISTENT, which is also the default value if no policy is specified. This is not a very good validation that the indexPolicy is actually being replaced.
I updated the test to replace the container with a different index policy (automatic = false and indexMode.NONE). This should be stronger validation

Comment (moderakh) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java]: thank you @Blackbaud-EricSlater 

@kushagraThapar please run the full CI.

Comment (kushagraThapar) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java]: Thanks @Blackbaud-EricSlater for updating the tests, however on the spring side, can you please add the support for CosmosReactiveRepository as well ?
Otherwise, when using reactiveRepositoies, the indexing policy update will not take affect.

Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/DocumentCollection.java]: will do

File: sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/domain/IndexPolicyEntity.java
Comment (kushagraThapar) [inline sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/domain/IndexPolicyEntity.java:1]: header missing.

File: sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/repository/integration/IndexPolicyUpdateIT.java
Comment (kushagraThapar) [inline sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/repository/integration/IndexPolicyUpdateIT.java:1]: header missing.

File: sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/repository/integration/ReactiveIndexPolicyUpdateIT.java
Comment (kushagraThapar) [inline sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/repository/integration/ReactiveIndexPolicyUpdateIT.java:1]: header missing.

**Review by kushagraThapar — CHANGES_REQUESTED** [review]
@Blackbaud-EricSlater - please add Microsoft headers on the new classes that are added in this PR. Otherwise looks good to me, thanks!

File: sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/repository/integration/IndexPolicyUpdateIT.java
Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/repository/integration/IndexPolicyUpdateIT.java:3]: fixed

File: sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/repository/integration/ReactiveIndexPolicyUpdateIT.java
Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/repository/integration/ReactiveIndexPolicyUpdateIT.java:3]: fixed

File: sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/domain/IndexPolicyEntity.java
Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-spring-data-cosmos-test/src/test/java/com/azure/spring/data/cosmos/domain/IndexPolicyEntity.java:3]: fixed

File: sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/SimpleCosmosRepository.java
Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/SimpleCosmosRepository.java]: @kushagraThapar any thoughts on this? The way i see it I either need to flesh out the comparison code to account for the discrepancies mentioned above or we omit the check entirely and let the container replace run everytime. Do you know what the cost of a container replace is? If it is an inexpensive constant time operation then letting it always run is probably fine. If it is something that is expensive and scales with the size of the container then we should ensure that needless container replaces are not being executed.

**Comment by Blackbaud-EricSlater** [comment]
@kushagraThapar on line 468 of CosmosTemplate, the createContainerIfNotExists method contains the following code...

`if (information.getRequestUnit() == null) {
    cosmosContainerResponseMono =
          cosmosAsyncDatabase.createContainerIfNotExists(cosmosContainerProperties);
} else {
    ThroughputProperties throughputProperties =
                        ThroughputProperties.createManualThroughput(information.getRequestUnit());
     cosmosContainerResponseMono = cosmosAsyncDatabase.createContainerIfNotExists(cosmosContainerProperties,
                            throughputProperties);
}`

Do i need to have similar code in place when doing the container replace? Do you now if the throughput properties are preserved or if they will be unset if i do a container replace without specifying them?

File: sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/SimpleCosmosRepository.java
Comment (kushagraThapar) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/SimpleCosmosRepository.java]: @Blackbaud-EricSlater - It is an expensive call that we should not do all the times. The current equals code that you have on this PR is costly or is it fine ?

**Comment by kushagraThapar** [comment]
> @kushagraThapar on line 468 of CosmosTemplate, the createContainerIfNotExists method contains the following code...
> 
> ```
> if (information.getRequestUnit() == null) {
> 		cosmosContainerResponseMono =
> 				cosmosAsyncDatabase.createContainerIfNotExists(cosmosContainerProperties);
> } else {
> 		ThroughputProperties throughputProperties =
> 				ThroughputProperties.createManualThroughput(information.getRequestUnit());
> 		cosmosContainerResponseMono =
> 				cosmosAsyncDatabase.createContainerIfNotExists(cosmosContainerProperties,
> 						throughputProperties);
> }
> ```
> 
> Do i need to have similar code in place when doing the container replace? I noticed that there is another replace method that accepts the throughput properties. Do you now if the throughput properties are preserved or if they will be unset if i do a container replace without specifying them?

We don't need to, because we are using the existing containerProperties and just updating the indexingPolicy information. It should preserve the throughput properties.

**Comment by kushagraThapar** [comment]
@Blackbaud-EricSlater - seems like the spring tests are failing, can you please fix it so I can run the full CI ?

File: sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/SimpleCosmosRepository.java
Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/SimpleCosmosRepository.java]: its not that the equals comparison is costly, its that it never returns true. I am working on fixing this now. I am adding definitions for .equals to the IncludedPath, ExcludedPath and Index classes and adding some additional code that accounts for the discrepancies i mentioned above.

Comment (kushagraThapar) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/SimpleCosmosRepository.java]: Makes sense, that will be the right way to fix this.

Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/SimpleCosmosRepository.java]: this is fixed now and the logic for comparing indexPolicies is contained in IndexPolicyCompareService. Implementing .equals on ExcludedPath and IncludedPath was necessary as well

**Comment by Blackbaud-EricSlater** [comment]
@kushagraThapar after talking with @Blackbaud-MikeLueders on my team, we were wondering if you guys knew the answer to the following...
If i update an index policy using the code as is, will the updated policy immediately be reflected if i retrieve the container properties right after? In my testing, the updated policy is immediately available but the container is empty. If the container were large with lots of documents, would there be a delay between me issuing the container replace and me retrieving the updated policy. 
If there is a delay then this could be problematic in cases where code is being executed by multiple instances of the same service, or in the event that the service fails to start, retries startup and executes the code again.

**Comment by kushagraThapar** [comment]
> @kushagraThapar after talking with @Blackbaud-MikeLueders on my team, we were wondering if you guys knew the answer to the following...
> If i update an index policy using the code as is, will the updated policy immediately be reflected if i retrieve the container properties right after? In my testing, the updated policy is immediately available but the container is empty. If the container were large with lots of documents, would there be a delay between me issuing the container replace and me retrieving the updated policy.
> If there is a delay then this could be problematic in cases where code is being executed by multiple instances of the same service, or in the event that the service fails to start, retries startup and executes the code again.

@moderakh  @anfeldma-ms to comment on this.

**Review by kushagraThapar — APPROVED** [review]

File: sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/IndexPolicyCompareService.java
Comment (kushagraThapar) [inline sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/IndexPolicyCompareService.java:19]: @Blackbaud-EricSlater - why do we need an instance here ?
Can't we static methods for this service ?
Since this is a stateless service, creating instance here is redundant in my opinion, thoughts ?

**Review by kushagraThapar — CHANGES_REQUESTED** [review]
Requested one change on instantiation of `IndexPolicyCompareService`

File: sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/IndexPolicyCompareService.java
Comment (kushagraThapar) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/IndexPolicyCompareService.java:19]: @Blackbaud-EricSlater - can you please implement the above changes so we can get this PR in ?
I am currently out on vacation, and won't be able to review after tomorrow. 
In case you are out too, we can merge it in January.

Comment (Blackbaud-EricSlater) [reply sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/repository/support/IndexPolicyCompareService.java]: fixed

**Comment by Blackbaud-EricSlater** [comment]
> @kushagraThapar after talking with @Blackbaud-MikeLueders on my team, we were wondering if you guys knew the answer to the following...
> If i update an index policy using the code as is, will the updated policy immediately be reflected if i retrieve the container properties right after? In my testing, the updated policy is immediately available but the container is empty. If the container were large with lots of documents, would there be a delay between me issuing the container replace and me retrieving the updated policy.
> If there is a delay then this could be problematic in cases where code is being executed by multiple instances of the same service, or in the event that the service fails to start, retries startup and executes the code again.

@kushagraThapar i do think we need an answer on this question before merging or the solution may contain issues when running in parallel.

**Comment by kushagraThapar** [comment]
> > @kushagraThapar after talking with @Blackbaud-MikeLueders on my team, we were wondering if you guys knew the answer to the following...
> > If i update an index policy using the code as is, will the updated policy immediately be reflected if i retrieve the container properties right after? In my testing, the updated policy is immediately available but the container is empty. If the container were large with lots of documents, would there be a delay between me issuing the container replace and me retrieving the updated policy.
> > If there is a delay then this could be problematic in cases where code is being executed by multiple instances of the same service, or in the event that the service fails to start, retries startup and executes the code again.
> 
> @kushagraThapar i do think we need an answer on this question before merging or the solution may contain issues when running in parallel.

@moderakh  - can you please answer this question ?
Its blocking the PR right now.

**Comment by moderakh** [comment]
> > > @kushagraThapar after talking with @Blackbaud-MikeLueders on my team, we were wondering if you guys knew the answer to the following...
> > > If i update an index policy using the code as is, will the updated policy immediately be reflected if i retrieve the container properties right after? In my testing, the updated policy is immediately available but the container is empty. If the container were large with lots of documents, would there be a delay between me issuing the container replace and me retrieving the updated policy.
> > > If there is a delay then this could be problematic in cases where code is being executed by multiple instances of the same service, or in the event that the service fails to start, retries startup and executes the code again.
> > 
> > 
> > @kushagraThapar i do think we need an answer on this question before merging or the solution may contain issues when running in parallel.
> 
> @moderakh - can you please answer this question ?
> Its blocking the PR right now.

@kushagraThapar  @Blackbaud-MikeLueders

container definition update:
If you update the container definition with a new indexing policy, upon cosmos db service acking the update the new indexing policy will be reflected in the container definition and any future fetching of container definition will have your update.

indexing in background:
after you update the container and service acks the container update, indexing the historical data will take time and that's something which will happen in background and how long it takes depends on how much data exists in the container.

**Comment by Blackbaud-EricSlater** [comment]
@kushagraThapar i am back from holiday time off. Is this good to merge? Looks like the answer to the previous question was good news and the solution should work as is.

**Comment by kushagraThapar** [comment]
> @kushagraThapar i am back from holiday time off. Is this good to merge? Looks like the answer to the previous question was good news and the solution should work as is.

@Blackbaud-EricSlater Sounds good, I will merge it today, thanks!

**Review by kushagraThapar — APPROVED** [review]

**Pull request merged**
